server/plugin.go

```diff
@@ -257,6 +259,8 @@ func (p *Plugin) getGitlabUserInfoByMattermostID(userID string) (*gitlab.UserInf
 
 	var userInfo gitlab.UserInfo
 
+	p.tokenLock.Lock()
+	defer p.tokenLock.Unlock()
```

Comment:
I'm not sure if we need this. @levb What do you think?

Reply:
This is required because when we click on the "Refresh" button in the UI, there are 4 API calls simultaneously and all of the handlers call this function at the same time in different goroutines. So, to prevent refreshing the token multiple times, this is required.

Reply:
@manojmalik20 We wouldn't want this lock to block for all users at the same time though. I'm wondering if we can do both of the following:

- Limit claiming the lock to a specific user
- Release the lock as soon as we know that the token is not expired

The token will be expired for a small percentage of API calls, so we should capitalize on this and avoid long claims of the lock when possible.

Thinking about it some more, I'm not sure if refreshing the token multiple times is an issue, considering how it would be a small percentage of API calls that will reach this case. Looking at the GitLab OAuth token docs I don't see any rate limiting on OAuth token refreshes so we're good there. Since it's only 4 API calls, I don't think we take a very big hit from refreshing for each of those calls in the worst case scenario.

Did you run into any issues when developing this without the lock? I'd say that we actually don't need this lock. @hanzei What are your thoughts on this?

Reply:
Okay, if refreshing the token multiple times is fine and we're willing to take that hit, I can remove the lock. Other than this, I had no issues without the lock.